File: eval/listenbrainz.py
"""ListenBrainz candidate-source client (T-C spike, eval harness only).

Adds a second candidate-generation path alongside Last.fm's artist.getsimilar:
labs.api.listenbrainz.org's similar-artists endpoint, which is MBID-keyed (not
name-keyed) and community-listening-session-based rather than tag/co-listen
based like Last.fm's. That makes it a structurally different candidate source
worth A/B'ing.

Two hosts are involved, each rate-limited independently to ~1 req/s with a
descriptive User-Agent (both labs.api.listenbrainz.org and musicbrainz.org ask
for this):

  1. labs.api.listenbrainz.org/similar-artists/json — the candidate source
     itself. Takes `artist_mbids` + a fixed `algorithm` enum, returns already
     name-attached neighbors (no reverse MBID->name lookup needed).
  2. musicbrainz.org/ws/2/artist — MBID resolution FALLBACK ONLY. Tier 1 is
     free: Last.fm's own artist.getinfo already returns an `mbid` field for
     most well-known artists, and that call is cached/shared with the rest of
     the pipeline (see `resolve_mbid`). MusicBrainz search is only hit when
     Last.fm has no mbid on file.

Resolution miss-rate (seed names that fail BOTH tiers, and so contribute zero
ListenBrainz candidates) is tracked on the client and surfaced in the harness
report — the spec calls it out as part of the answer, not an implementation
detail: a lot of the "does ListenBrainz beat Last.fm" verdict lives here.

Every network call on both hosts is sqlite-cached like the Last.fm client,
in eval/.cache/listenbrainz.sqlite (a separate file so a cold ListenBrainz run
never writes into the pre-warmed Last.fm cache).
"""
from __future__ import annotations
import asyncio
import json
import pathlib
import time
import urllib.parse

import aiohttp

from cache import Cache

ROOT = pathlib.Path(__file__).resolve().parent
LB_CACHE_PATH = ROOT / ".cache" / "listenbrainz.sqlite"

LB_SIMILAR_URL = "https://labs.api.listenbrainz.org/similar-artists/json"
MB_SEARCH_URL = "https://musicbrainz.org/ws/2/artist/"
# Fixed algorithm enum (labs API rejects anything else) — long window (7500
# sessions), moderate contribution/threshold floor, pre-filtered, limit 100.
LB_ALGORITHM = (
    "session_based_days_7500_session_300_contribution_5_threshold_10_"
    "limit_100_filter_True_skip_30"
)
USER_AGENT = (
    "obscurity-engine-eval-spike/0.1 "
    "( T-C ListenBrainz candidate-source A/B; contact: gauravg@deepnative.ai )"
)


class _RateLimiter:
    """Serializes calls to ~1 per `interval` seconds for one host.

    A lock-protected timestamp, not a token bucket: every caller awaits its
    turn, so concurrent asyncio.gather() callers still can't exceed the rate,
    regardless of how many users/seeds are in flight at once.
    """

    def __init__(self, interval: float = 1.05):
        self.interval = interval
        self._lock = asyncio.Lock()
        self._last = 0.0

    async def wait(self) -> None:
        async with self._lock:
            now = time.monotonic()
            delta = now - self._last
            if delta < self.interval:
                await asyncio.sleep(self.interval - delta)
            self._last = time.monotonic()


class ListenBrainzClient:
    def __init__(self, cache: Cache | None = None):
        self.cache = cache or Cache(LB_CACHE_PATH)
        self.session: aiohttp.ClientSession | None = None
        self._lb_limiter = _RateLimiter(1.05)
        self._mb_limiter = _RateLimiter(1.05)
        self._mbid_memo: dict[str, str | None] = {}  # in-process, avoids duplicate races
        self._mbid_locks: dict[str, asyncio.Lock] = {}
        # resolution diagnostics — the "miss-rate" the spec asks for
        self.stat_resolved_lastfm = 0
        self.stat_resolved_musicbrainz = 0
        self.stat_unresolved = 0
        self.stat_similar_calls = 0
        self.stat_similar_empty = 0

    async def start(self) -> None:
        if self.session is None:
            timeout = aiohttp.ClientTimeout(total=15, connect=5)
            self.session = aiohttp.ClientSession(
                timeout=timeout, headers={"User-Agent": USER_AGENT}
            )

    async def close(self) -> None:
        if self.session:
            await self.session.close()
            self.session = None

    # ── low-level cached GET (shared by both hosts, keyed by prefix) ──────────

    async def _get(self, url: str, params: dict, limiter: _RateLimiter, prefix: str):
        ck = prefix + ":" + urllib.parse.urlencode(dict(sorted(params.items())))
        cached = self.cache.get(ck)
        if cached is not None:
            try:
                return json.loads(cached)
            except ValueError:
                return None
        await limiter.wait()
        try:
            async with self.session.get(url, params=params) as resp:
                if resp.status != 200:
                    return None
                text = await resp.text()
                try:
                    data = json.loads(text)
                except ValueError:
                    return None
                self.cache.set(ck, text)
                return data
        except Exception:  # noqa: BLE001 - network resilience, mirrors lastfm.py
            return None

    # ── MBID resolution: Last.fm mbid field (free) -> MusicBrainz search ──────

    async def resolve_mbid(self, name: str, lastfm_client) -> str | None:
        """Resolve an artist display name to a MusicBrainz ID.

        Tier 1: Last.fm's artist.getinfo `mbid` field — reuses the Last.fm
        client's own cache/key-rotation, so this is free when that info was
        (or will be) fetched anyway, and still cheap (a single Last.fm call,
        not rate-limited beyond Last.fm's own pool) when it wasn't.
        Tier 2: MusicBrainz artist search, rate-limited ~1 req/s, cached.
        """
        if name in self._mbid_memo:
            return self._mbid_memo[name]
        lock = self._mbid_locks.setdefault(name, asyncio.Lock())
        async with lock:
            if name in self._mbid_memo:  # resolved while we waited for the lock
                return self._mbid_memo[name]
            mbid = await self._resolve_via_lastfm(name, lastfm_client)
            if mbid:
                self.stat_resolved_lastfm += 1
            else:
                mbid = await self._resolve_via_musicbrainz(name)
                if mbid:
                    self.stat_resolved_musicbrainz += 1
                else:
                    self.stat_unresolved += 1
            self._mbid_memo[name] = mbid
            return mbid

    async def _resolve_via_lastfm(self, name: str, lastfm_client) -> str | None:
        # Uses the Last.fm client's own cached _get — no new network policy,
        # no double-fetch: if the pipeline already needed this artist's info
        # for scoring, this is a pure cache read.
        data = await lastfm_client._get({"method": "artist.getinfo", "artist": name})
        artist = data.get("artist") if isinstance(data, dict) else None
        if not artist:
            return None
        mbid = artist.get("mbid")
        return mbid or None

    async def _resolve_via_musicbrainz(self, name: str) -> str | None:
        query = f'artist:"{name}"'
        data = await self._get(
            MB_SEARCH_URL,
            {"query": query, "fmt": "json", "limit": 3},
            self._mb_limiter,
            "mb:search",
        )
        if not data:
            return None
        artists = data.get("artists") or []
        if not artists:
            return None
        # highest-score result only; MB search is already relevance-ranked
        best = artists[0]
        if best.get("score", 0) < 80:  # weak match — don't attach the wrong artist
            return None
        return best.get("id")

    # ── similar-artists ─────────────────────────────────────────────────────

    async def similar(self, mbid: str, limit: int) -> list[tuple[str, float]]:
        data = await self._get(
            LB_SIMILAR_URL,
            {"artist_mbids": mbid, "algorithm": LB_ALGORITHM},
            self._lb_limiter,
            "lb:similar",
        )
        self.stat_similar_calls += 1
        if not isinstance(data, list) or not data:
            self.stat_similar_empty += 1
            return []
        # LB `score` is an unbounded raw count (community listening-session
        # co-occurrence), not a 0-1 match — normalize per-seed so it composes
        # with the rest of the pipeline's conviction math the same way
        # Last.fm's already-normalized `match` does.
        max_score = max((float(a.get("score", 0) or 0) for a in data), default=0.0) or 1.0
        out = []
        for a in sorted(data, key=lambda a: a.get("score", 0), reverse=True)[:limit]:
            name = a.get("name")
            if not name:
                continue
            match = float(a.get("score", 0) or 0) / max_score
            out.append((name, match))
        return out

    def resolution_stats(self) -> dict:
        total = (
            self.stat_resolved_lastfm
            + self.stat_resolved_musicbrainz
            + self.stat_unresolved
        )
        return {
            "resolved_via_lastfm": self.stat_resolved_lastfm,
            "resolved_via_musicbrainz": self.stat_resolved_musicbrainz,
            "unresolved": self.stat_unresolved,
            "total_resolution_attempts": total,
            "miss_rate": (self.stat_unresolved / total) if total else 0.0,
            "similar_calls": self.stat_similar_calls,
            "similar_empty": self.stat_similar_empty,
        }


# ── process-wide singleton ─────────────────────────────────────────────────
# Shared across every concurrent run_user() call in the harness process so the
# two host rate limiters are actually global (per-user instances would let N
# concurrent users each fire at ~1/s => N req/s in aggregate — not polite).

_singleton: ListenBrainzClient | None = None
_singleton_lock = asyncio.Lock()


async def get_client() -> ListenBrainzClient:
    global _singleton
    async with _singleton_lock:
        if _singleton is None:
            _singleton = ListenBrainzClient()
            await _singleton.start()
        return _singleton


async def close_client() -> None:
    global _singleton
    if _singleton is not None:
        await _singleton.close()
        _singleton = None


# ── candidate-map construction (same shape as pipeline.build_candidates) ───

async def build_candidates_lb(lastfm_client, names: list[str], cfg) -> dict:
    """norm -> {"display": str, "recs": {seed_name: match}, "hop": 1}

    Mirrors pipeline.build_candidates's output shape exactly so the rest of
    the pipeline (scoring, diversity, tier variety) is candidate-source
    agnostic. Seeds that fail MBID resolution simply contribute no candidates
    — that's the reach cost of the miss-rate, not a bug.
    """
    from pipeline import normalize_artist_name  # deferred: avoid import cycle

    client = await get_client()
    mbids = await asyncio.gather(*(client.resolve_mbid(n, lastfm_client) for n in names))
    pairs = [(seed, mbid) for seed, mbid in zip(names, mbids) if mbid]
    sims = await asyncio.gather(*(client.similar(mbid, cfg.similar_per_seed) for _s, mbid in pairs))

    cmap: dict[str, dict] = {}
    for (seed, _mbid), neighbors in zip(pairs, sims):
        for cand_name, match in neighbors:
            norm = normalize_artist_name(cand_name)
            entry = cmap.setdefault(norm, {"display": cand_name, "recs": {}, "hop": 1})
            entry["recs"][seed] = max(entry["recs"].get(seed, 0.0), match)
    return cmap


def merge_candidate_maps(lf_map: dict, lb_map: dict) -> dict:
    """Union two candidate maps (already keyed by normalized name), taking the
    max per-seed match when both sources recommend the same candidate off the
    same seed (mirrors the existing 1-hop/2-hop merge rule in pipeline.py)."""
    merged: dict[str, dict] = {norm: {"display": e["display"], "recs": dict(e["recs"]), "hop": e["hop"]}
                                for norm, e in lf_map.items()}
    for norm, entry in lb_map.items():
        target = merged.setdefault(norm, {"display": entry["display"], "recs": {}, "hop": entry["hop"]})
        for seed, match in entry["recs"].items():
            target["recs"][seed] = max(target["recs"].get(seed, 0.0), match)
    return merged

File: backend/src/listenbrainz.rs
//! ListenBrainz candidate-source client — the additive second discovery graph.
//!
//! Mirrors `eval/listenbrainz.py` (the reference implementation that the SHIP
//! verdict `docs/blend-n348-2026-07-03.md` validated): a second candidate path
//! alongside Last.fm's `artist.getsimilar`, sourced from
//! `labs.api.listenbrainz.org`'s community-listening-session similar-artists
//! endpoint. It is MBID-keyed (not name-keyed), so names are resolved to
//! MusicBrainz IDs in two tiers before the graph can be queried:
//!
//!   1. Last.fm's own `artist.getinfo` `mbid` field (free — reuses the pool).
//!   2. MusicBrainz `/ws/2/artist` search, accepted only at score >= 80.
//!
//! ## Fail-open (deliberately different from Last.fm's fail-closed rule)
//! The Last.fm candidate path is fail-CLOSED: a transient failure aborts the
//! whole request so the pool stays deterministic (pool determinism drives the
//! cache key + persisted run). ListenBrainz is purely ADDITIVE, so it is
//! fail-OPEN: if LB is slow or down, the request degrades to Last.fm-only
//! candidates and never errors or meaningfully delays a discovery. The 8s
//! per-request time budget (see `candidates.rs`) enforces this: whatever LB
//! answered within budget is blended, the rest is skipped and the request is
//! counted as degraded. NEVER convert an LB failure into a request failure.
//! The budget runs CONCURRENTLY with the Last.fm arm (the two arms race — see
//! `candidates.rs::race_arms`), overlapping LB's window with Last.fm's work
//! rather than following it; a Last.fm-arm error drops the in-flight LB future
//! in place (it is never spawned, so cancellation leaks nothing).
//!
//! ## Caching (mandatory before the lever is flipped)
//! Every similar-artists and MBID resolution is cached through the shared
//! `CacheStore` (in-memory default, Redis when `REDIS_URL` is set) with a LONG
//! 7-day TTL — LB similarity is stable. Both hits AND definitive misses
//! (unresolvable names, empty neighbour lists) are cached so misses don't
//! re-fetch. Keys are user-agnostic (`lb:mbid:<norm>`, `lb:similar:<mbid>`).
//!
//! ## Politeness
//! The labs API has no key auth — do not hammer it. Bounded concurrency (4)
//! plus a small jittered spacing on the similar endpoint; MusicBrainz (which
//! explicitly asks for ~1 req/s) is serialized behind a 1s pacer. A descriptive
//! User-Agent is sent on both hosts (both ask for one).

use std::sync::atomic::{AtomicBool, Ordering};
use std::time::{Duration, Instant};

use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tokio::sync::{Mutex, Semaphore};

use crate::cache::CacheStore;
use crate::lastfm::LastfmClient;
use crate::metrics::Metrics;
use crate::utils::normalize_artist_name;

const LB_SIMILAR_URL: &str = "https://labs.api.listenbrainz.org/similar-artists/json";
const MB_SEARCH_URL: &str = "https://musicbrainz.org/ws/2/artist/";

/// Fixed algorithm enum (the labs API rejects anything else) — long window,
/// moderate contribution/threshold floor, pre-filtered, limit 100. Identical to
/// the eval's `LB_ALGORITHM`, so the Rust engine queries the same graph the
/// n=348 SHIP verdict measured.
const LB_ALGORITHM: &str = "session_based_days_7500_session_300_contribution_5_threshold_10_limit_100_filter_True_skip_30";

const USER_AGENT: &str = concat!(
    "obscurity-engine/",
    env!("CARGO_PKG_VERSION"),
    " ( ListenBrainz blend candidate source; contact: gauravg@deepnative.ai )"
);

/// 7 days — LB similarity is stable, so cache aggressively (the verdict doc's
/// own caveat: caching is mandatory before the lever can be flipped in prod).
pub const LB_CACHE_TTL: Duration = Duration::from_secs(7 * 24 * 3600);

/// Bounded concurrency for the (no-auth) labs similar-artists endpoint.
const LB_SIMILAR_CONCURRENCY: usize = 4;
/// MusicBrainz asks for ~1 req/s — serialize and pace it.
const MB_MIN_INTERVAL: Duration = Duration::from_millis(1050);
/// MusicBrainz search score below this is a weak match — don't attach the wrong
/// artist (mirrors the eval's threshold).
const MB_SCORE_FLOOR: i64 = 80;

/// The candidate-generation source, selected by the `CANDIDATE_SOURCE` env var.
/// Default (and any unset/invalid value) is `Lastfm` — exactly today's behavior.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CandidateSource {
    /// Today's behavior: Last.fm `artist.getsimilar` only.
    Lastfm,
    /// ListenBrainz similar-artists only (structurally different graph).
    ListenBrainz,
    /// Union of both sources (the SHIP-verified `blend`).
    Blend,
}

impl CandidateSource {
    /// Parse `CANDIDATE_SOURCE`. Unset/blank/invalid → `Lastfm` (the safe default
    /// — the lever is flipped in prod only after review, never implicitly).
    pub fn from_env() -> Self {
        match std::env::var("CANDIDATE_SOURCE") {
            Ok(v) => Self::parse(&v),
            Err(_) => Self::Lastfm,
        }
    }

    pub fn parse(raw: &str) -> Self {
        match raw.trim().to_ascii_lowercase().as_str() {
            "listenbrainz" => Self::ListenBrainz,
            "blend" => Self::Blend,
            _ => Self::Lastfm, // "lastfm", empty, and anything unrecognized
        }
    }

    /// Whether this source needs a ListenBrainz client wired up.
    pub fn uses_listenbrainz(self) -> bool {
        matches!(self, Self::ListenBrainz | Self::Blend)
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Lastfm => "lastfm",
            Self::ListenBrainz => "listenbrainz",
            Self::Blend => "blend",
        }
    }
}

/// A minimum-interval pacer: serializes callers so at most one request per
/// `min_interval` goes out on a host (used for MusicBrainz). Lock-protected
/// timestamp, not a token bucket — every caller awaits its turn.
struct Pacer {
    min_interval: Duration,
    last: Mutex<Option<Instant>>,
}

impl Pacer {
    fn new(min_interval: Duration) -> Self {
        Self { min_interval, last: Mutex::new(None) }
    }

    async fn wait(&self) {
        let mut guard = self.last.lock().await;
        if let Some(last) = *guard {
            let elapsed = last.elapsed();
            if elapsed < self.min_interval {
                tokio::time::sleep(self.min_interval - elapsed).await;
            }
        }
        *guard = Some(Instant::now());
    }
}

/// Cached MBID resolution. `mbid: None` is a *negative* cache entry (the name is
/// unresolvable via both tiers) — cached so we don't re-attempt it every run.
#[derive(Serialize, Deserialize)]
struct MbidCache {
    mbid: Option<String>,
}

/// Cached similar-artists list (already normalized + score-sorted). An empty
/// `neighbors` is a valid negative-cache entry (mbid genuinely has no LB
/// neighbours), so it is not re-fetched.
#[derive(Serialize, Deserialize)]
struct SimilarCache {
    neighbors: Vec<(String, f64)>,
}

pub struct ListenBrainzClient {
    http: Client,
    /// Bounds concurrent calls to the no-auth labs similar endpoint.
    sim_sem: Semaphore,
    /// Serializes MusicBrainz to one caller at a time (paired with `mb_pacer`).
    mb_sem: Semaphore,
    mb_pacer: Pacer,
    /// Liveness signal for `/api/status`: set true on any successful LB fetch,
    /// false when a similar-artists call fails at the transport/5xx level. A 404
    /// (unknown mbid) is a legitimate miss and does NOT flip this. Starts true so
    /// a fresh process with no traffic reads "ok".
    healthy: AtomicBool,
    /// Endpoints — the real hosts in production; injectable so unit tests can
    /// point at a local mock server (no live ListenBrainz/MusicBrainz in tests).
    similar_url: String,
    mb_url: String,
}

impl Default for ListenBrainzClient {
    fn default() -> Self {
        Self::new()
    }
}

impl ListenBrainzClient {
    pub fn new() -> Self {
        Self::with_urls(LB_SIMILAR_URL.to_string(), MB_SEARCH_URL.to_string())
    }

    /// Construct with explicit endpoints (production uses `new`; tests point
    /// these at a local mock server).
    pub fn with_urls(similar_url: String, mb_url: String) -> Self {
        let http = Client::builder()
            .timeout(Duration::from_secs(10))
            .connect_timeout(Duration::from_secs(5))
            .user_agent(USER_AGENT)
            .build()
            .unwrap_or_default();
        Self {
            http,
            sim_sem: Semaphore::new(LB_SIMILAR_CONCURRENCY),
            mb_sem: Semaphore::new(1),
            mb_pacer: Pacer::new(MB_MIN_INTERVAL),
            healthy: AtomicBool::new(true),
            similar_url,
            mb_url,
        }
    }

    /// For `/api/status`. True when the last LB fetch succeeded (or none yet).
    pub fn healthy(&self) -> bool {
        self.healthy.load(Ordering::Relaxed)
    }

    /// Test-only: force the health flag (to exercise the `/api/status` "error"
    /// state without needing a real transport failure).
    #[cfg(test)]
    pub fn mark_unhealthy(&self) {
        self.healthy.store(false, Ordering::Relaxed);
    }

    // ── MBID resolution: Last.fm mbid (free) → MusicBrainz search ────────────

    /// Resolve an artist display name to a MusicBrainz ID, two-tier and cached
    /// (positive AND negative). Cache key uses `normalize_artist_name` so the
    /// dedup semantics match the rest of the pipeline.
    pub async fn resolve_mbid(
        &self,
        name: &str,
        lastfm: &LastfmClient,
        cache: &CacheStore,
        metrics: &Metrics,
    ) -> Option<String> {
        let key = format!("lb:mbid:{}", normalize_artist_name(name));
        if let Some(c) = cache.get_json::<MbidCache>(&key).await {
            metrics.record_lb_cache_hit();
            return c.mbid;
        }
        // Tier 1: Last.fm's own mbid field — a pool call (often already cached),
        // no new rate policy. Empty string counts as "no mbid".
        let mut mbid = lastfm
            .fetch_artist_mbid(name)
            .await
            .ok()
            .flatten()
            .filter(|m| !m.is_empty());
        // Tier 2: MusicBrainz search (rate-limited), only if tier 1 missed.
        if mbid.is_none() {
            mbid = self.mb_search(name).await;
        }
        // Cache the outcome either way (negative cache prevents re-attempts).
        cache.put_json(&key, &MbidCache { mbid: mbid.clone() }, LB_CACHE_TTL).await;
        mbid
    }

    async fn mb_search(&self, name: &str) -> Option<String> {
        let _permit = self.mb_sem.acquire().await.ok()?;
        self.mb_pacer.wait().await;
        let query = format!("artist:\"{}\"", name);
        let resp = match self
            .http
            .get(&self.mb_url)
            .query(&[("query", query.as_str()), ("fmt", "json"), ("limit", "3")])
            .send()
            .await
        {
            Ok(r) => r,
            Err(_) => return None, // network error → treat as unresolved (fail-open)
        };
        if !resp.status().is_success() {
            return None;
        }
        let v: Value = resp.json().await.ok()?;
        let best = v.get("artists")?.as_array()?.first()?;
        // MB search is relevance-ranked; take the top result only if strong.
        let score = best
            .get("score")
            .and_then(|s| s.as_i64().or_else(|| s.as_str().and_then(|x| x.parse().ok())))
            .unwrap_or(0);
        if score < MB_SCORE_FLOOR {
            return None;
        }
        best.get("id").and_then(|i| i.as_str()).map(str::to_string)
    }

    // ── similar-artists ──────────────────────────────────────────────────────

    /// Fetch (cached) the LB similar-artists neighbourhood for one MBID, as
    /// `(display_name, normalized_match)` pairs. LB's raw `score` is an unbounded
    /// listening-session co-occurrence count, so it is normalized per-seed by the
    /// max so it composes with the pipeline's conviction math the same way
    /// Last.fm's already-0..1 `match` does. Infallible: any failure returns `[]`
    /// (fail-open) — but transport/5xx failures also flip the health flag and are
    /// NOT cached (so a transient outage doesn't poison the cache with an empty).
    pub async fn similar(
        &self,
        mbid: &str,
        limit: usize,
        cache: &CacheStore,
        metrics: &Metrics,
    ) -> Vec<(String, f64)> {
        let key = format!("lb:similar:{}", mbid);
        if let Some(c) = cache.get_json::<SimilarCache>(&key).await {
            metrics.record_lb_cache_hit();
            return c.neighbors.into_iter().take(limit).collect();
        }

        let _permit = match self.sim_sem.acquire().await {
            Ok(p) => p,
            Err(_) => return Vec::new(),
        };
        // Small jittered spacing so a burst of permit-holders doesn't fire in
        // lockstep against the no-auth labs endpoint.
        let jitter = 50 + (rand::random::<u64>() % 150);
        tokio::time::sleep(Duration::from_millis(jitter)).await;

        let resp = match self
            .http
            .get(&self.similar_url)
            .query(&[("artist_mbids", mbid), ("algorithm", LB_ALGORITHM)])
            .send()
            .await
        {
            Ok(r) => r,
            Err(_) => {
                self.healthy.store(false, Ordering::Relaxed);
                return Vec::new();
            }
        };
        let status = resp.status();
        if !status.is_success() {
            if status == reqwest::StatusCode::NOT_FOUND {
                // Unknown mbid — a stable, legitimate miss: negative-cache it.
                cache
                    .put_json(&key, &SimilarCache { neighbors: Vec::new() }, LB_CACHE_TTL)
                    .await;
            } else {
                // 5xx / 429 / etc — transient. Don't cache; mark unhealthy.
                self.healthy.store(false, Ordering::Relaxed);
            }
            return Vec::new();
        }
        let v: Value = match resp.json().await {
            Ok(v) => v,
            Err(_) => return Vec::new(),
        };
        self.healthy.store(true, Ordering::Relaxed);

        let neighbors = normalize_similar(&v);
        // Cache the full normalized+sorted list (empty included = negative cache).
        cache.put_json(&key, &SimilarCache { neighbors: neighbors.clone() }, LB_CACHE_TTL).await;
        neighbors.into_iter().take(limit).collect()
    }
}

/// Parse the labs similar-artists JSON array into normalized, score-sorted
/// `(name, match)` pairs. Pulled out as a pure fn so it is unit-testable without
/// a network. Mirrors `eval/listenbrainz.py`'s `similar()` normalization.
fn normalize_similar(v: &Value) -> Vec<(String, f64)> {
    let Some(arr) = v.as_array() else {
        return Vec::new();
    };
    let score_of = |a: &Value| -> f64 {
        a.get("score")
            .and_then(|s| s.as_f64().or_else(|| s.as_str().and_then(|x| x.parse().ok())))
            .unwrap_or(0.0)
    };
    let max_score = arr.iter().map(score_of).fold(0.0_f64, f64::max);
    let max_score = if max_score > 0.0 { max_score } else { 1.0 };
    let mut scored: Vec<(&Value, f64)> = arr.iter().map(|a| (a, score_of(a))).collect();
    scored.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    scored
        .into_iter()
        .filter_map(|(a, raw)| {
            a.get("name")
                .and_then(|n| n.as_str())
                .map(|name| (name.to_string(), raw / max_score))
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn candidate_source_parse_matrix() {
        assert_eq!(CandidateSource::parse("lastfm"), CandidateSource::Lastfm);
        assert_eq!(CandidateSource::parse("listenbrainz"), CandidateSource::ListenBrainz);
        assert_eq!(CandidateSource::parse("blend"), CandidateSource::Blend);
        // case / whitespace insensitivity
        assert_eq!(CandidateSource::parse("  BLEND "), CandidateSource::Blend);
        assert_eq!(CandidateSource::parse("ListenBrainz"), CandidateSource::ListenBrainz);
        // unset/invalid → the safe default, exactly today's behavior
        assert_eq!(CandidateSource::parse(""), CandidateSource::Lastfm);
        assert_eq!(CandidateSource::parse("nonsense"), CandidateSource::Lastfm);
        assert_eq!(CandidateSource::parse("LASTFM"), CandidateSource::Lastfm);
        assert!(!CandidateSource::Lastfm.uses_listenbrainz());
        assert!(CandidateSource::ListenBrainz.uses_listenbrainz());
        assert!(CandidateSource::Blend.uses_listenbrainz());
    }

    #[test]
    fn normalize_similar_scales_by_max_and_sorts() {
        let v = serde_json::json!([
            { "name": "Beta", "score": 5 },
            { "name": "Alpha", "score": 10 },
            { "name": "Gamma", "score": 0 },
        ]);
        let out = normalize_similar(&v);
        assert_eq!(out[0], ("Alpha".to_string(), 1.0)); // top = max, normalized to 1.0
        assert_eq!(out[1], ("Beta".to_string(), 0.5));
        assert_eq!(out[2], ("Gamma".to_string(), 0.0));
    }

    #[test]
    fn normalize_similar_handles_string_scores_and_missing_names() {
        let v = serde_json::json!([
            { "name": "A", "score": "20" },
            { "score": 40 },              // no name → dropped
            { "name": "B", "score": "10" },
        ]);
        let out = normalize_similar(&v);
        // Two named entries survive; normalized by global max (40).
        assert_eq!(out.len(), 2);
        assert_eq!(out[0], ("A".to_string(), 0.5));
        assert_eq!(out[1], ("B".to_string(), 0.25));
    }

    #[test]
    fn normalize_similar_empty_and_nonarray() {
        assert!(normalize_similar(&serde_json::json!([])).is_empty());
        assert!(normalize_similar(&serde_json::json!({ "not": "an array" })).is_empty());
    }

    // ── Mock-server tests: cache hit/miss/negative, fail-open, MB score gate ──
    // No live ListenBrainz/MusicBrainz — a local axum server stands in.

    use crate::cache::{CacheStore, InMemoryStore};
    use crate::metrics::Metrics;
    use axum::{extract::{Query, State}, routing::get, Json, Router};
    use std::collections::HashMap;
    use std::sync::atomic::AtomicUsize;
    use std::sync::Arc;

    fn mem_cache() -> CacheStore {
        CacheStore::InMemory(InMemoryStore::new())
    }

    /// Spin a mock similar-artists server on an ephemeral port. It counts the
    /// requests it actually serves (so tests can prove a second call was a cache
    /// hit) and returns canned neighbours keyed by the `artist_mbids` query.
    async fn spawn_similar_mock() -> (String, Arc<AtomicUsize>) {
        let hits = Arc::new(AtomicUsize::new(0));
        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        let app = Router::new()
            .route(
                "/",
                get(|Query(q): Query<HashMap<String, String>>, State(h): State<Arc<AtomicUsize>>| async move {
                    h.fetch_add(1, Ordering::Relaxed);
                    let mbid = q.get("artist_mbids").cloned().unwrap_or_default();
                    if mbid == "empty" {
                        Json(serde_json::json!([]))
                    } else {
                        Json(serde_json::json!([
                            { "name": "Alpha", "score": 10 },
                            { "name": "Beta", "score": 5 },
                        ]))
                    }
                }),
            )
            .with_state(Arc::clone(&hits));
        tokio::spawn(async move { axum::serve(listener, app).await.unwrap(); });
        (format!("http://{}/", addr), hits)
    }

    #[tokio::test]
    async fn similar_cache_hit_miss_and_negative_cache() {
        let (url, hits) = spawn_similar_mock().await;
        let client = ListenBrainzClient::with_urls(url, "unused".into());
        let cache = mem_cache();
        let metrics = Metrics::new();

        // Miss → fetches, normalizes, caches. Server sees exactly one request.
        let first = client.similar("mbid1", 20, &cache, &metrics).await;
        assert_eq!(first, vec![("Alpha".into(), 1.0), ("Beta".into(), 0.5)]);
        assert_eq!(hits.load(Ordering::Relaxed), 1);
        assert_eq!(metrics.snapshot().lb_cache_hits, 0);

        // Hit → identical result, NO new server request, cache-hit counted.
        let second = client.similar("mbid1", 20, &cache, &metrics).await;
        assert_eq!(second, first);
        assert_eq!(hits.load(Ordering::Relaxed), 1, "second call served from cache");
        assert_eq!(metrics.snapshot().lb_cache_hits, 1);

        // Negative cache: an empty neighbour list is cached, not re-fetched.
        let empty1 = client.similar("empty", 20, &cache, &metrics).await;
        assert!(empty1.is_empty());
        assert_eq!(hits.load(Ordering::Relaxed), 2);
        let empty2 = client.similar("empty", 20, &cache, &metrics).await;
        assert!(empty2.is_empty());
        assert_eq!(hits.load(Ordering::Relaxed), 2, "empty result negative-cached");
        assert_eq!(metrics.snapshot().lb_cache_hits, 2);
    }

    #[tokio::test]
    async fn similar_fail_open_on_transport_error() {
        // Port 9 (discard) is expected closed → connection refused fast. The
        // client must degrade to [] and flip its health flag, never hang/panic.
        let client = ListenBrainzClient::with_urls("http://127.0.0.1:9/".into(), "unused".into());
        let cache = mem_cache();
        let metrics = Metrics::new();
        assert!(client.healthy(), "starts healthy");
        let out = client.similar("mbid1", 20, &cache, &metrics).await;
        assert!(out.is_empty(), "transport error degrades to no candidates (fail-open)");
        assert!(!client.healthy(), "transport failure flips health to error");
    }

    #[tokio::test]
    async fn mb_search_accepts_strong_rejects_weak() {
        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        let app = Router::new().route(
            "/",
            get(|Query(q): Query<HashMap<String, String>>| async move {
                let query = q.get("query").cloned().unwrap_or_default();
                // Weak match for names containing "weak", strong otherwise.
                let score = if query.to_lowercase().contains("weak") { 50 } else { 95 };
                Json(serde_json::json!({ "artists": [ { "id": "resolved-id", "score": score } ] }))
            }),
        );
        tokio::spawn(async move { axum::serve(listener, app).await.unwrap(); });
        let url = format!("http://{}/", addr);
        let client = ListenBrainzClient::with_urls("unused".into(), url);

        assert_eq!(client.mb_search("StrongMatch").await, Some("resolved-id".into()));
        assert_eq!(client.mb_search("WeakMatch").await, None, "score < 80 rejected");
    }
}
